openshift-kube-apiserver/admission/customresourcevalidation/dns/validate_dns.go

```diff
@@ -0,0 +1,203 @@
+type dnsV1 struct{}
+
+// ValidateCreate validates a DNS that is being created.
+func (dnsV1) ValidateCreate(uncastObj runtime.Object) field.ErrorList {
+	obj, errs := toDNSV1(uncastObj)
+	if len(errs) > 0 {
+		return errs
+	}
+
+	errs = append(errs, validation.ValidateObjectMeta(&obj.ObjectMeta, false, validation.NameIsDNSSubdomain, field.NewPath("metadata"))...)
+	errs = append(errs, validateDNSSpecCreate(obj.Spec)...)
+
+	return errs
+}
+
+// ValidateUpdate validates a DNS that is being updated.
+func (dnsV1) ValidateUpdate(uncastObj runtime.Object, uncastOldObj runtime.Object) field.ErrorList {
+	obj, errs := toDNSV1(uncastObj)
+	if len(errs) > 0 {
+		return errs
+	}
+	oldObj, errs := toDNSV1(uncastOldObj)
+	if len(errs) > 0 {
+		return errs
+	}
+
+	errs = append(errs, validation.ValidateObjectMetaUpdate(&obj.ObjectMeta, &oldObj.ObjectMeta, field.NewPath("metadata"))...)
+	errs = append(errs, validateDNSSpecUpdate(obj.Spec, oldObj.Spec)...)
+
+	return errs
+}
+
+// ValidateStatusUpdate validates a DNS status that is being updated.
+func (dnsV1) ValidateStatusUpdate(uncastObj runtime.Object, uncastOldObj runtime.Object) field.ErrorList {
+	obj, errs := toDNSV1(uncastObj)
+	if len(errs) > 0 {
+		return errs
+	}
+	oldObj, errs := toDNSV1(uncastOldObj)
+	if len(errs) > 0 {
+		return errs
+	}
+
+	errs = append(errs, validation.ValidateObjectMetaUpdate(&obj.ObjectMeta, &oldObj.ObjectMeta, field.NewPath("metadata"))...)
+
+	return errs
+}
+
+// validateDNSSpecCreate validates the spec of a DNS that is being created.
+func validateDNSSpecCreate(spec operatorv1.DNSSpec) field.ErrorList {
+	return validateDNSSpec(spec)
+}
+
+// validateDNSSpecUpdate validates the spec of a DNS that is being updated.
+func validateDNSSpecUpdate(newspec, oldspec operatorv1.DNSSpec) field.ErrorList {
+	return validateDNSSpec(newspec)
+}
+
+// validateDNSSpec validates the spec of a DNS.
+func validateDNSSpec(spec operatorv1.DNSSpec) field.ErrorList {
+	var errs field.ErrorList
+	specField := field.NewPath("spec")
+	errs = append(errs, validateDNSNodePlacement(spec.NodePlacement, specField.Child("nodePlacement"))...)
+	return errs
+}
+
+// validateDNSSpec validates the spec.nodePlacement field of a DNS.
+func validateDNSNodePlacement(nodePlacement operatorv1.DNSNodePlacement, fldPath *field.Path) field.ErrorList {
+	var errs field.ErrorList
+	if len(nodePlacement.NodeSelector) != 0 {
+		errs = append(errs, unversionedvalidation.ValidateLabels(nodePlacement.NodeSelector, fldPath.Child("nodeSelector"))...)
+	}
+	if len(nodePlacement.Tolerations) != 0 {
+		errs = append(errs, validateTolerations(nodePlacement.Tolerations, fldPath.Child("tolerations"))...)
+	}
+	return errs
+}
+
+// validateTolerations is a copy of
+// "k8s.io/kubernetes/pkg/apis/core/validation".ValidateTolerations that has
+// been modified to validate a versioned Toleration.
+func validateTolerations(tolerations []corev1.Toleration, fldPath *field.Path) field.ErrorList {
```

Comment:
Ordinarily its not allowed, but because this actually lives in k/k `Convert_v1_Toleration_To_core_Toleration` is usable.  Sorry I didn't think of it before.  That also keeps your validation correct over time.

Reply:
Done.  Thanks!  That simplifies things nicely.